File: Java Practice/ProgrammingPractice2021/src/LeetcodeProblems/ThreeSumClosest.java
package LeetcodeProblems;

import java.util.Arrays;

public class ThreeSumClosest {

    public static int threeSumClosest(int[] nums, int target) {
        int minDiff = Integer.MAX_VALUE;
        int closestSum = 0;
        Arrays.sort(nums);
        for(int i = 0 ; i <= nums.length -3; i++){
            if(i == 0 || nums[i] > nums[i-1]){
                int start = i + 1;
                int end = nums.length - 1;
                while(start < end){
                    int currSum = nums[i] + nums[start] + nums[end];
                    int diff = Math.abs(currSum - target);
                    if(currSum == target) return target;
                    if(diff < minDiff){
                        minDiff = diff;
                        closestSum = currSum;
                    }
                    if(currSum < target){
                        int currStart = start;
                        while(nums[currStart] == nums[start] && start < end){
                            start++;
                        }
                    }else{
                        int currEnd = end;
                        while(nums[currEnd] == nums[end] && start < end){
                            end--;
                        }
                    }
                }
            }
        }
        return closestSum;
    }

    public static void main(String[] args) {
        int inputNums[] = {-1, 2, 1, -4};
        int target = 1;
        System.out.println(threeSumClosest(inputNums, target));
    }

}

File: C++ Practice/Programming/ProgrammingProblems/ThreeSumClosest.cpp
/**
 * Given an integer array nums of length n and an integer target, find three integers in nums such that the sum is closest to target.
 * Return the sum of the three integers.
 * You may assume that each input would have exactly one solution.
 */
#include<iostream>
#include<vector>
using namespace std;
    int threeSumClosest(vector<int> &nums, int target){
        vector<vector<int>> resultSet;
        int minClosest = INT_MAX;
        int closestSum = 0;
        if(nums.size() > 2){
            sort(nums.begin(), nums.end());
            for(int i = 0 ; i <= nums.size() - 3 ; i++) {
                if (i == 0 || nums[i] > nums[i - 1]) {
                    int start = i + 1;
                    int end = nums.size() - 1;
                    while (start < end) {
                        int currSum = nums[i] + nums[start] + nums[end];
                        if(currSum == target) return currSum;
                        int diff = abs(currSum - target);
                        if (diff < minClosest) {
                            minClosest = diff;
                            closestSum = currSum;
                        }
                        if (currSum < target) {
                            int currStart = start;
                            while (nums[currStart] == nums[start] && start < end) {
                                start++;
                            }
                        } else {
                            int currEnd = end;
                            while (nums[end] == nums[currEnd] && start < end) {
                                end--;
                            }
                        }
                    }
                }
            }
        }
        return closestSum;
    }
int main(){
        vector<int> inputNum = {-1,2,1,-4};
        int result = threeSumClosest(inputNum, 1);
        cout<< result << endl;
        return result;
}


